draft/XXXX-table-metadata/README.md

```diff
@@ -0,0 +1,80 @@
+
+2. Providing this information to other applications, so that they do not have to do this data inspection themselves, or if this information is required by those applications. Specifically, [UpSet](https://github.com/VCG/upset) is the first application for which this will be used.
+
+## Proposals
+
+#### Data Model
+There are several ways we could go about this, but in my opinion, the cleanest way is to store this information in a special collection, in the workspace that the table resides in. This special collection would be named something like `multinet_meta`, and each entry would correspond to the metadata on a specific graph or table. I think the data model would look something like this:
+
+
+```
+{
+  item_id: str                 // The ID of the table/graph this data belongs to
+  item_type: "table" | "graph" // The major type of the item
+  table?: {                    // Only present if the major type is "table"
+    type: "edge" | "node"      // The table type
+    columns: [
+      {
+        key: str               // E.g. {"foo": "bar"}, "foo" is the key
+        type: "number" | "label" | "category" | "date" | "bool"
+      }
+    ]
+  }
+  graph?: {                    // Only present if the major type is "graph"
+    // TBD
+  }
+
+}
+```
+
+#### API
+There are two main operations that are necessary for this functionality.
+
+1. Specifying the metadata on table creation
+2. Retrieving this metadata
+
+Ideally, our API would allow for the creation of a table/graph record, with more versatile options for specifying metadata. However, our API isn't currently structured this way. Instead, in order to conform to our existing API, we would likely just add this metadata object as a query parameter onto each upload endpoint. For retrieving this metadata, the API should resemble the following:
+
+```
+GET /api/workspaces/{workspace}/tables/{table}/metadata
+```
+
+##### Specifying Metadata on Table Creation
+
+The data model of the API would differ slightly from that of what we store internally, to allow for a single request to describe metadata on multiple entities (mainly in the case of graphs). The data format of this API may resemble the following:
+
+```
+{
+  graph?: {
+    // TBD, any relevant graph metadata
+  }
+  tables: [  // For a single CSV, this array must contain only one item.
+    {
+      // Unchanged from the table field above
+    }
+  ]
+}
+```
+
+However, handling the case of graphs may not be necessary in the first approach. Feedback on how to best implement this is welcome.
+
+## Backwards Compatibility
+
+All changes may be backwards compatible to the server, as this new query parameter would ideally be an "optional" parameter. However, if it's desired that this field be mandatory, then this breaks backwards compatibility for all upload endpoints.
```

Comment:
I think this should be mandatory, because some visualizations will work best when it's specified. I'd like to hear some counterarguments, though, if you guys have some

Reply:
I largely agree with @AlmightyYakob on this. But de facto it will be mandatory in a sense: since we expect the vast majority of people to upload data to Multinet through the client, then whatever the client does will appear to be the normal course of action. We already have the type specification UI built for CSV uploads; this means that when all the pieces are connected, _all_ CSV uploads will include type information. Maybe in some future revision we can decide to make this mandatory, but I don't see a need for it now.

If there's no type information for a particular column, I think it would make sense to implicitly treat that column as `label` (since this is effectively the current semantics of typeless upload anyway).

Reply:
Yeah, I'd say that for any missing columns, it is assumed to be label, and just stored as string. This would also apply for all metadata, so if no metadata is passed along, all columns are just treated as strings/labels.

Reply:
You can edit this section to say that there won't be any breakages.

Reply:
b2409bf